File: ios/clarity_flutter/Sources/clarity_flutter/bridges/DeviceBridge.swift
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import Foundation
import UIKit

internal struct DeviceBridge {

    private static let bytesPerMb: UInt64 = 1_048_576

    static func getDeviceInfo() -> DeviceInfoData {
        #if targetEnvironment(simulator)
        let machine = ProcessInfo.processInfo.environment["SIMULATOR_MODEL_IDENTIFIER"] ?? "unknown"
        #else
        var systemInfo = utsname()
        uname(&systemInfo)
        let machine = withUnsafePointer(to: &systemInfo.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(cString: $0)
            }
        }
        #endif

        let totalMemoryMb = Int64(clamping: ProcessInfo.processInfo.physicalMemory / bytesPerMb)
        return DeviceInfoData(brand: nil, machine: machine, totalMemory: totalMemoryMb)
    }

    static func getPackageInfo() -> PackageInfoData {
        let bundle = Bundle.main
        let infoDictionary = bundle.infoDictionary ?? [:]
        return PackageInfoData(
            packageName: bundle.bundleIdentifier ?? "",
            version: (infoDictionary["CFBundleShortVersionString"] as? String) ?? "",
            buildNumber: (infoDictionary["CFBundleVersion"] as? String) ?? ""
        )
    }

    static func getCacheDirectory() -> String? {
        return FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?.path
    }

    static func getUserAgent() -> String {
        let device = UIDevice.current
        let osString = device.systemVersion.replacingOccurrences(of: ".", with: "_")
        return "Mozilla/5.0 (\(device.model); CPU iPhone OS \(osString) like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148"
    }
}

File: android/src/main/kotlin/com/microsoft/clarity/bridges/DeviceBridge.kt
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package com.microsoft.clarity.bridges

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import android.webkit.WebSettings
import com.microsoft.clarity.generated.DeviceInfoData
import com.microsoft.clarity.generated.PackageInfoData

internal object DeviceBridge {
    private const val BYTES_PER_MB = 1_048_576L

    fun getDeviceInfo(context: Context): DeviceInfoData =
        try {
            val activityManager =
                context.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
            val totalMemoryMb =
                if (activityManager != null) {
                    val memoryInfo = ActivityManager.MemoryInfo()
                    activityManager.getMemoryInfo(memoryInfo)
                    memoryInfo.totalMem / BYTES_PER_MB
                } else {
                    0L
                }
            DeviceInfoData(brand = Build.BRAND, machine = null, totalMemory = totalMemoryMb)
        } catch (_: Throwable) {
            DeviceInfoData(brand = null, machine = null, totalMemory = 0L)
        }

    fun getPackageInfo(context: Context): PackageInfoData =
        try {
            val packageName = context.packageName
            val info = context.packageManager.getPackageInfo(packageName, 0)
            val buildNumber =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    info.longVersionCode.toString()
                } else {
                    @Suppress("DEPRECATION")
                    info.versionCode.toString()
                }
            PackageInfoData(
                packageName = packageName,
                version = info.versionName ?: "",
                buildNumber = buildNumber,
            )
        } catch (_: Throwable) {
            PackageInfoData(packageName = "", version = "", buildNumber = "")
        }

    fun getCacheDirectory(context: Context): String? =
        try {
            context.cacheDir.path
        } catch (_: Throwable) {
            null
        }

    fun getUserAgent(context: Context): String =
        try {
            WebSettings.getDefaultUserAgent(context)
        } catch (_: Throwable) {
            ""
        }
}
